### PR TITLE
[RVV] Optimize Generic RVV Matmul codegen

### PR DESCRIPTION
**Pull request by bhbruce**
### Background
RVV matmul always uses register tile size [m, n] = [8, 32]. It's ok for VLEN=512. It can generate outer product GEMM with LMUL=2. The vector register utilization rate is 56%(18/32). However, it uses LMUL=1 for VLEN=1024 and the vector register utilization rate is 28%. If we use VLEN = 128. The codegen uses LMUL=8 and vector registers are not enough. It introduces register spill and causes performance drop.
### Optimization
Two optimizations are included in this PR to improve RVV matmul code generation.

1. Register tiling optimization: We introduce VLEN-aware register tiling. The register tile size is changed from [m, n, k] = [8, 32, 1] to [7, vlmax_lmul_4, 1]. This can not only fix the problem of generating inefficient code for `VLEN!=512` but also increase the vector register utilization rate to 100%. Based on experiments on SiFive platforms, the performance improvement is around 25%~50%.

2. Cache tiling optimization: 
  - After introducing VLEN-aware register tiling, the original cache tiling code may break since `maxTileSize` is set to 64 by default. To solve the problem, we guarantee `maxTileSize >= minTileSize`, which implies that cache tile size >= register tile size.
  -   Introducing option `iree-llvmcpu-aggressive-distribution` makes cache tile size become twice of register tile size if `maxTileSize < minTileSize`. It reduces cache miss rate and leads to performance improvement. Performance improvement is 3~5% in general.

In summary, my experiment on the SiFive VLEN 1024 platform shows x5.4 speedup for Bert-Large after applying the patches.

### CODE REVIEW

**Comment by bhbruce** [comment]
FYI. @rednoah91

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 1]
```diff
@@ -61,6 +61,14 @@ static llvm::cl::opt<int>
...
                           llvm::cl::desc("default distribution tile size"),
                           llvm::cl::init(64));
+static llvm::cl::opt<bool> clEnableAggressiveDist(
+    "iree-llvmcpu-aggressive-distribution",
+    llvm::cl::desc(
+        "Enable aggressive method for distribution tile size. "
+        "It is only applied for linalg contraction ops now. "
+        "If distConfig.minTileSizes[i] >= distConfig.maxTileSizes[i], "
+        "set distConfig.maxTileSizes[i] to 2 * distConfig.minTileSizes[i]."),
+    llvm::cl::init(false));
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Is this for risc-v only? Can we add `-riscv-` to the flag for now? I don't see a plan that we'll use it for all the targets at this moment.

Also, could you move it to line 108 (i.e., below `clDisableArmSMETiling`), so we can group all the flags related to CPU targets in the same place?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 2]
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
...
+  int64_t nativeVectorSize = getNativeVectorSizeInBytes(entryPointFn);
+  int64_t elementSize;
+  if (elementType->isF16()) {
+    elementSize = 16;
+  } else if (elementType->isF32()) {
+    elementSize = 32;
+  } else if (elementType->isF64()) {
+    elementSize = 64;
+  } else {
+    // ToDo: support int data type
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: style nit: TODO

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 3]
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
...
+///
+/// TODO: Currently it only supports for nonWideningLinalgElementType.
+static void
+getMatmulRISCVVectorSizes(mlir::FunctionOpInterface entryPointFn,
+                          linalg::LinalgOp op, int64_t vectorSize,
+                          SmallVectorImpl<int64_t> &sizes,
+                          SmallVectorImpl<bool> &scalableSizeFlags) {
+  if (sizes.empty())
+    getDefaultMatmulVectorSizes(op, vectorSize, sizes, scalableSizeFlags);
+  // Todo: support widening matmul.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: style nit: TODO

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 4]
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
...
+  // Determines n dimension tile size with VLEN for
+  // nonWideningLinalgElementType.
+  auto elementType = nonWideningLinalgElementType(op);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Can you spell out the type? It is `FailureOr<Type>`.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 5]
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
...
+  } else {
+    // ToDo: support int data type
+    return;
+  }
+  Value lhs = op.getDpsInputs()[0];
+  ShapedType lhsType = cast<ShapedType>(lhs.getType());
+  auto lhsShape = lhsType.getShape();
+  bool has_batch = isa<linalg::BatchMatmulOp, linalg::BatchMatmulTransposeBOp>(
+      op.getOperation());
+  char idx = (has_batch) ? 1 : 0;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: nit: use `int`. Should we use it in all those `sizes` access (e.g., should below all become `sizes[baseIdx+0/1/2/3/4]`? If so, I think we can rename it to `baseIdx`.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 6]
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
...
+  } else if (elementType->isF32()) {
+    elementSize = 32;
+  } else if (elementType->isF64()) {
+    elementSize = 64;
+  } else {
+    // ToDo: support int data type
+    return;
+  }
+  Value lhs = op.getDpsInputs()[0];
+  ShapedType lhsType = cast<ShapedType>(lhs.getType());
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: style nit: we follow mixed LLVM and Google coding style in IREE. We should use `auto` here because `cast` already spells the type.

https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 7]
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
...
+  auto lhsShape = lhsType.getShape();
+  bool has_batch = isa<linalg::BatchMatmulOp, linalg::BatchMatmulTransposeBOp>(
+      op.getOperation());
+  char idx = (has_batch) ? 1 : 0;
+
+  // Use 7 x lmul4 to fully utilize vector registers.
+  sizes[0] = 7;
+  // Calculate tile size for the main vector dimension (N).
+  constexpr int64_t byteSizeInBits = 8;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: style nit

https://google.github.io/styleguide/cppguide.html#Constant_Names

```suggestion
  constexpr int64_t kByteSizeInBits = 8;
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 8]
```diff
@@ -1354,6 +1420,18 @@ getMatmulVectorSizes(mlir::FunctionOpInterface entryPointFn,
     }
   }
 
+  if (isRISCV(targetAttr) && hasAnyVFeature(targetAttr)) {
+    bool isMTB = isa<linalg::MatmulTransposeBOp>(op);
+    bool isBMTB = isa<linalg::BatchMatmulTransposeBOp>(op);
+    // Use default tile size for matmul_transpose_b &
+    // batch_matmul_transpose_b to avoid performance drop.
+    if (!isMTB && !isBMTB) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: nit: you can collapse it to one `isa` in the if condition.

```suggestion
    // Use default tile size for matmul_transpose_b &
    // batch_matmul_transpose_b to avoid performance drop.
    if (!isa<linalg::MatmulTransposeBOp, linalg::BatchMatmulTransposeBOp>(op)) {
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir:1]: What is the purpose of adding this test? IMO, it really looks like a test with transform spec, although you're using C++ passes. It is not really maintainable in the IREE main repo because it is testing a combination of passes. The lowering_config is preset, and the TileAndFuse tests are covered in the other file.

If we want to do the pipeline test, are you able to add a test in [pipeline_tests.mlir](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir)? Or create a `pipeline_aggressive_distribution_riscv_tests.mlir` and add the test to the new file? (Note, you'll need to use `builtin.module(iree-llvmcpu-select-lowering-strategy, func.func(iree-llvmcpu-lower-executable-target))` but not these arbitrary passes.)

The main difference is that the pipeline tests run the full pipeline which includes strategy selection and the lowering. This gives us a better picture about what we're expecting in the test.

**Review by hanhanW — CHANGES_REQUESTED** [review]
Thanks for the contribution! I left some comments, please take a look.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 6 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: ok.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 7 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: solved.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 8 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Applied suggestion.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 4 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Thank you for catching it. Let me modify it.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 1 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Yes, it's for RVV only. I modify it to `iree-llvmcpu-riscv-aggressive-distribution` and move it to line 108.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 5 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: This `idx` aims to point to `m dimension index` in `lhsShape`.
How about changing to `mDimIdx`?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir:1]: > What is the purpose of adding this test?

I would like to ensure that the compiler can help generate the desired for-loop struct if using register tile size [7, 64] and cache tile size [64,128].

I agree with you. I should create the other file to run the whole pipeline test. Thank you for providing the details.

**Comment by bhbruce** [comment]
Hi @hanhanW 
Thank you for your suggestions. I've followed your reviewed comments to modify it.
I appended several commits at the end of the original commits.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 5 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I see. In this case, I think what you're looking for is [inferContractionDims](https://github.com/llvm/llvm-project/blob/15d1560ea4047a2b4b14c826767089f538ddda70/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h#L50-L65). This provides a more flexible way to not just look at named ops. You can get the batch/m/n/k dimensions from it.

Furthermore, you can bail out if one of the dimension is greater than one.

EDIT: This is an optional review comment. I don't know your end goal, but I'd like to share that there is an utility for inferring these dimensions.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 9]
```diff
@@ -1289,6 +1298,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
...
+    elementSize = 32;
+  } else if (elementType->isF64()) {
+    elementSize = 64;
+  } else {
+    // TODO: support int data type
+    return;
+  }
+  Value lhs = op.getDpsInputs()[0];
+  auto lhsType = cast<ShapedType>(lhs.getType());
+  auto lhsShape = lhsType.getShape();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I did not notice that what you need is just the shape. How about using some linalg interface method to get it? You can do something like `ArrayRef<int64_t> lhsShape = linalgOp.getShape(linalgOp.getDpsInputOperand(0));`. Then we don't need the `lhs` and `lhsType` variables.

https://github.com/llvm/llvm-project/blob/1f2509993e6e0717b547b5214b06550af4f3008f/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L417-L439

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir:1]: I think we can collapse this file into the other one. What you're looking for is adding another `// RUN...` with the flag. E.g.,

https://github.com/iree-org/iree/blob/c651ba9257315af96e75a371537f3cf8ab080cb0/compiler/src/iree/compiler/Codegen/Common/test/decompose_pack_unpack_ops.mlir#L1-L2

This way reduces the dup tests in different files.

**Review by hanhanW — COMMENTED** [review]
Overall looks good to me, just few nits.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir:1]: OK. I'll collapse it to `compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir`.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 9 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I'm new to MLIR. Thank you for the info.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 5 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Thank you for providing this approach. However, I am confused in why each dimension has size two.

**Comment by bhbruce** [comment]
Hi @hanhanW 
Based on your review comments, I've updated the PR.
Please help to review.
Thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir:1]: E.g., here is a contraction op but the number of M dimensions is more than 1.

```mlir
util.func public @multi_m_dim_generic(%arg0 : tensor<64x4x128xf32>, %arg1 : tensor<128x512xf32>,
    %arg2 : tensor<64x4x512xf32>) -> tensor<64x4x512xf32> {
    %4 = linalg.generic {
        indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                         affine_map<(d0, d1, d2, d3) -> (d2, d1)>,
                         affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>],
        iterator_types = ["parallel", "parallel", "parallel", "reduction"]}
        ins(%arg0, %arg1 : tensor<64x4x128xf32>, tensor<128x512xf32>) outs(%arg2 : tensor<64x4x512xf32>) {
    ^bb0(%in: f32, %in_0: f32, %out: f32):
      %5 = arith.mulf %in, %in_0 : f32
      %6 = arith.addf %5, %out : f32
      linalg.yield %6 : f32
    } -> tensor<64x4x512xf32>
  util.return %4 : tensor<64x4x512xf32>
}
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir:1]: Oops, sorry for not being clear enough. What I meant is collapsing the `select_riscv_lowering_strategy_*.mlir` to the `select_riscv_lowering_strategy.mlir` file, not the pipeline_test one.

The original `pipeline_riscv_aggressive_distribution_tests.mlir` looks good to me, can you recover it?

**Comment by hanhanW** [comment]
> Hi @hanhanW Based on your review comments, I've updated the PR. Please help to review. Thanks!

I left the comment about how we structure the tests. Other parts look okay to me.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir:1]: oh... My bad... I misunderstood it. Let me revert and merge `select_riscv_lowering_strategy_*.mlir`.

Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir:1]: Solved!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir:1]: Do we face this kind of MLIR in simple pass routing?
I supposed that multi_m & multi_n only happens when enabling mmt4d.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir [hunk 10]
```diff
@@ -30,6 +31,105 @@ func.func @matmul_riscv() attributes {hal.executable.target = #executable_target
...
+
+#pipeline_layout = #hal.pipeline.layout<bindings = [
+  #hal.pipeline.binding<storage_buffer>,
+  #hal.pipeline.binding<storage_buffer>,
+  #hal.pipeline.binding<storage_buffer>
+]>
+
+#executable_target_embedded_elf_riscv_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-riscv_64", {cpu_features = "+m,+a,+f,+d,+zvl1024b,+v", data_layout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128", native_vector_size = 256 : index, target_triple = "riscv64-unknown-unknown-eabi-elf"}>
+builtin.module {
+  func.func @matmul_riscv_vl1024() attributes {hal.executable.target = #executable_target_embedded_elf_riscv_64_} {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir]: What is the difference between this `matmul_riscv_vl1024` test and the below `matmul_riscv_vl1024` test? Can we collapse them into one? (In this case, you can move all your `CHECK-AGGRESSIVE`s right after the regular `CHECK`s). E.g.,

```
module {....}
// CHECK: ...
// CHECK: ...
// CHECK: ...
...

// CHECK-AGGRESSIVE: ...
// CHECK-AGGRESSIVE: ...
// CHECK-AGGRESSIVE: ...
...

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir [hunk 11]
```diff
@@ -30,6 +31,105 @@ func.func @matmul_riscv() attributes {hal.executable.target = #executable_target
...
+
+#pipeline_layout = #hal.pipeline.layout<bindings = [
+  #hal.pipeline.binding<storage_buffer>,
+  #hal.pipeline.binding<storage_buffer>,
+  #hal.pipeline.binding<storage_buffer>
+]>
+
+#executable_target_embedded_elf_riscv_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-riscv_64", {cpu_features = "+m,+a,+f,+d,+zvl512b,+v", data_layout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128", native_vector_size = 128 : index, target_triple = "riscv64-unknown-unknown-eabi-elf"}>
+builtin.module {
+  func.func @gemv_riscv_vl512() attributes {hal.executable.target = #executable_target_embedded_elf_riscv_64_} {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir]: Same question, I can't distinguish the difference between this and the above one based on the name. What is the difference?

**Review by hanhanW — COMMENTED** [review]
Wh

EDIT: I did not mean to submit this comment, please ignore this one.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir (hunk 11 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir]: This is different in problem size. LHS is 1 by 512, so it is GEMMV. 
Other tests are GEMM test.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir (hunk 10 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir]: The matmul problem sizes were slightly different between them. 
I merged two MLIR files into one to avoid confusion.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir:1]: I don't have handy IR/examples now. However, it is not mmt4d specific. We've seen that in some models. One of design points of the data-tiling (e.g., mmt4d things) is that we can narrow down the variants for GEMM kernels.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir (hunk 11 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir]: I see, can you put the "gemmv" and "gemm" words to the function name in this case?

**Review by hanhanW — COMMENTED** [review]
LGTM with the final nit: https://github.com/iree-org/iree/pull/18986#discussion_r1847163822

Thanks for addressing my comments!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir:1]: OK, thank you for clarifying. I'll keep in mind for this.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir (hunk 11 above)
Comment (bhbruce) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir]: Ok. I also fix the ci error in previous commits. Let's wait for the result.

**Comment by bhbruce** [comment]
Thank you @hanhanW. I learn a lot from your reviewed comments.

**Review by hanhanW — APPROVED** [review]

**Comment by hanhanW** [comment]
There is a failure on risc-v side, could you take a look?

```
[184/554] Generating check_llvm-cpu_local-task_dot_general.mlir_module.vmfb from dot_general.mlir
FAILED: tests/e2e/stablehlo_ops/check_llvm-cpu_local-task_dot_general.mlir_module.vmfb /home/runner/work/iree/iree/build-riscv/tests/e2e/stablehlo_ops/check_llvm-cpu_local-task_dot_general.mlir_module.vmfb 
cd /home/runner/work/iree/iree/build-riscv/tests/e2e/stablehlo_ops && /home/runner/work/iree/iree/.venv/bin/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=llvm-cpu --iree-input-type=stablehlo --iree-input-demote-f64-to-f32 --iree-llvmcpu-target-cpu=generic /home/runner/work/iree/iree/tests/e2e/stablehlo_ops/dot_general.mlir -o check_llvm-cpu_local-task_dot_general.mlir_module.vmfb --iree-hal-executable-object-search-path=\"/home/runner/work/iree/iree/build-riscv\" --iree-llvmcpu-target-triple=riscv64 --iree-llvmcpu-target-abi=lp64d --iree-llvmcpu-target-cpu-features=+m,+a,+f,+d,+c,+zvl512b,+v --riscv-v-fixed-length-vector-lmul-max=8
Please report issues to https://github.com/iree-org/iree/issues and include the crash backtrace.
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  libIREECompiler.so 0x00007ff69f24e6f8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 40
1  libIREECompiler.so 0x00007ff69f24c49e llvm::sys::RunSignalHandlers() + 238
2  libIREECompiler.so 0x00007ff69f24edad
3  libpthread.so.0    0x00007ff699d35420
4  libIREECompiler.so 0x00007ff6a0e6ad35
5  libIREECompiler.so 0x00007ff6a0e585e9
6  libIREECompiler.so 0x00007ff6a0e5827c
7  libIREECompiler.so 0x00007ff6a0e57dc6
8  libIREECompiler.so 0x00007ff69f464dd9 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 633
9  libIREECompiler.so 0x00007ff69f465bb9 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 329
10 libIREECompiler.so 0x00007ff69f4679d8 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) + 4712
11 libIREECompiler.so 0x00007ff69f4652a8 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 1864
12 libIREECompiler.so 0x00007ff69f465bb9 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 329
13 libIREECompiler.so 0x00007ff69f46ad2a
14 libIREECompiler.so 0x00007ff6a0a2b698
15 libIREECompiler.so 0x00007ff69f464dd9 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 633
16 libIREECompiler.so 0x00007ff69f465bb9 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 329
17 libIREECompiler.so 0x00007ff69f4679d8 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) + 4712
18 libIREECompiler.so 0x00007ff69f4652a8 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 1864
19 libIREECompiler.so 0x00007ff69f465bb9 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 329
20 libIREECompiler.so 0x00007ff69f46ad2a
21 libIREECompiler.so 0x00007ff6a0a2c239
22 libIREECompiler.so 0x00007ff69f464dd9 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 633
23 libIREECompiler.so 0x00007ff69f465bb9 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 329
24 libIREECompiler.so 0x00007ff69f46b8d2
25 libIREECompiler.so 0x00007ff69f3548f8
26 libIREECompiler.so 0x00007ff69f354857
27 libpthread.so.0    0x00007ff699d324df
28 libIREECompiler.so 0x00007ff69f354c33
29 libIREECompiler.so 0x00007ff69f354cc5
30 libIREECompiler.so 0x00007ff69f207f88 llvm::StdThreadPool::processTasks(llvm::ThreadPoolTaskGroup*) + 1416
31 libIREECompiler.so 0x00007ff69f2093ae
32 libpthread.so.0    0x00007ff699d29609
33 libc.so.6          0x00007ff699c4e353 clone + 67
```

**Comment by bhbruce** [comment]
The error happens when input is `linalg.vecmat`.
In that case, `cDims->m.size()` is 0. I add a guard check if `cDims->m.size() != 1` to return. Because we only optimize for matmul now instead of all contraction ops.
